drivers/serial/uart_emul.c

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright (c) 2023 Fabian Blatz
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT zephyr_uart_emul
+
+#include <zephyr/kernel.h>
+#include <zephyr/drivers/uart.h>
+#include <zephyr/sys/ring_buffer.h>
+#include <zephyr/drivers/serial/uart_emul.h>
+#include <errno.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(uart_emul, CONFIG_UART_LOG_LEVEL);
+
+struct uart_emul_config {
+	bool loopback;
+};
+
+/* Device run time data */
+struct uart_emul_data {
+	struct uart_config cfg;
+
+	struct ring_buf *rx_rb;
+	struct k_spinlock rx_lock;
+
+	uart_emul_callback_tx_data_ready_t tx_data_ready_cb;
+	struct ring_buf *tx_rb;
+	struct k_spinlock tx_lock;
+};
+
+int uart_emul_poll_in(const struct device *dev, unsigned char *p_char)
+{
+	struct uart_emul_data *drv_data = dev->data;
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&drv_data->rx_lock);
+
+	if (ring_buf_is_empty(drv_data->rx_rb)) {
+		LOG_DBG("Rx buffer is empty");
+		k_spin_unlock(&drv_data->rx_lock, key);
+		return -ENOBUFS;
```

Comment:
If there is no data, this needs to return `-1` as described in `uart_poll_in()`.

Reply:
Ah, didn't know that, sorry @faxe1008 for my earlier comment then.

Reply:
All good, reverted it :^)